File: BackEnd/Api/Domain/Entities/VesselType.cs
namespace Api.Domain.Entities;

using Api.Domain.ValueObjects;
using Api.Application.DataTransfer;
using Api.Infrastructure.Utilities;

public class VesselType : IDTOAble<VesselTypeDto>
{
    public Guid Id { get; private set; }

    public Designation Name { get; private set; }
    public Designation Description { get; private set; }
    public uint MaxNumberOfRows { get; private set; }
    public uint MaxNumberOfBays { get; private set; }
    public uint MaxNumberOfTiers { get; private set; }
    public PhysicalCharacteristics PhysicalCharacteristics { get; private set; }
    public virtual ICollection<Dock> Docks { get; private set; }
    public virtual ICollection<Vessel> Vessels { get; private set; }

    //EF Core
    protected VesselType() { }
    public VesselType(Guid id, Designation name, Designation description, uint maxNumberOfRows, uint maxNumberOfBays, uint maxNumberOfTiers, PhysicalCharacteristics physicalCharacteristics)
    {

        if (maxNumberOfBays == 0) throw new ArgumentException("Max number of bays must be greater than zero.");
        if (maxNumberOfRows == 0) throw new ArgumentException("Max number of rows must be greater than zero.");
        if (maxNumberOfTiers == 0) throw new ArgumentException("Max number of tiers must be greater than zero.");

        Id = id;
        Name = name ?? throw new ArgumentNullException(nameof(name));
        Description = description ?? throw new ArgumentNullException(nameof(description));
        MaxNumberOfRows = maxNumberOfRows;
        MaxNumberOfBays = maxNumberOfBays;
        MaxNumberOfTiers = maxNumberOfTiers;
        PhysicalCharacteristics = physicalCharacteristics ?? throw new ArgumentNullException(nameof(physicalCharacteristics));
    }

    public VesselTypeDto ToDTO()
    {
        return new VesselTypeDto
        {
            Name = this.Name.Value,
            Description = this.Description.Value,
            MaxNumberOfRows = this.MaxNumberOfRows,
            MaxNumberOfBays = this.MaxNumberOfBays,
            MaxNumberOfTiers = this.MaxNumberOfTiers,
            PhysicalCharacteristics = this.PhysicalCharacteristics
        };
    }

    public void UpdateName(string name)
    {
        Name = new Designation { Value = name };
    }
    public void UpdateDescription(string description)
    {
        Description = new Designation { Value = description };
    }
    public void UpdateMaxNumberOfRows(uint maxNumberOfRows)
    {
        if (maxNumberOfRows == 0) throw new ArgumentException("Max number of rows must be greater than zero.");
        MaxNumberOfRows = maxNumberOfRows;
    }
    public void UpdateMaxNumberOfBays(uint maxNumberOfBays)
    {
        if (maxNumberOfBays == 0) throw new ArgumentException("Max number of bays must be greater than zero.");
        MaxNumberOfBays = maxNumberOfBays;
    }
    public void UpdateMaxNumberOfTiers(uint maxNumberOfTiers)
    {
        if (maxNumberOfTiers == 0) throw new ArgumentException("Max number of tiers must be greater than zero.");
        MaxNumberOfTiers = maxNumberOfTiers;
    }
    public void UpdatePhysicalCharacteristics(PhysicalCharacteristics physicalCharacteristics)
    {
        if (Vessels != null && Vessels.Count > 0)
        {
            foreach (var v in Vessels)
            {
                if (v.PhysicalCharacteristics.Length > physicalCharacteristics.Length ||
                   v.PhysicalCharacteristics.Depth > physicalCharacteristics.Depth ||
                   v.PhysicalCharacteristics.Draft > physicalCharacteristics.Draft)
                {
                    throw new InvalidOperationException("Cannot update physical characteristics of a vessel type assigned to vessels with greater physical characteristics.");
                }
            }
        }

        if (Docks != null && Docks.Count > 0)
        {
            foreach (var d in Docks)
            {
                if (d.PhysicalCharacteristics.Length < physicalCharacteristics.Length ||
                   d.PhysicalCharacteristics.Depth < physicalCharacteristics.Depth ||
                   d.PhysicalCharacteristics.Draft < physicalCharacteristics.Draft)
                {
                    throw new InvalidOperationException("Cannot update physical characteristics of a vessel type assigned to docks with lesser physical characteristics.");
                }
            }
        }
        PhysicalCharacteristics = physicalCharacteristics ?? throw new ArgumentNullException(nameof(physicalCharacteristics));
    }
}

File: FrontEnd/src/model/VesselType.ts
import type { VesselTypeDto } from './dto/VesselTypeDto';
import type { PhysicalCharacteristics } from './values/PhysicalCharacteristics';

export class VesselType {
    
    private _name: string;
    private _description: string;
    private _maxNumberOfRows: number;
    private _maxNumberOfBays: number;
    private _maxNumberOfTiers: number;
    private _physicalCharacteristics: PhysicalCharacteristics;

    constructor(params: {
        name: string;
        description: string;
        maxNumberOfRows: number;
        maxNumberOfBays: number;
        maxNumberOfTiers: number;
        capacity?: number;
        physicalCharacteristics: PhysicalCharacteristics;
    }) {
        if (!params.name) throw new Error('Name cannot be null or empty.');
        if (!params.description) throw new Error('Description cannot be null or empty.');
        if (!params.physicalCharacteristics) throw new Error('Physical characteristics cannot be null.');
        if (params.maxNumberOfBays === 0) throw new Error('Max number of bays must be greater than zero.');
        if (params.maxNumberOfRows === 0) throw new Error('Max number of rows must be greater than zero.');
        if (params.maxNumberOfTiers === 0) throw new Error('Max number of tiers must be greater than zero.');

        this._name = params.name;
        this._description = params.description;
        this._maxNumberOfRows = params.maxNumberOfRows;
        this._maxNumberOfBays = params.maxNumberOfBays;
        this._maxNumberOfTiers = params.maxNumberOfTiers;
        this._physicalCharacteristics = params.physicalCharacteristics;
    }

    get name(): string { return this._name; }
    get description(): string { return this._description; }
    get maxNumberOfRows(): number { return this._maxNumberOfRows; }
    get maxNumberOfBays(): number { return this._maxNumberOfBays; }
    get maxNumberOfTiers(): number { return this._maxNumberOfTiers; }
    get capacity(): number { return this._maxNumberOfRows * this._maxNumberOfBays * this._maxNumberOfTiers; }
    get physicalCharacteristics(): PhysicalCharacteristics { return this._physicalCharacteristics; }

    updateName(name: string): void {
        if (!name) throw new Error('Name cannot be null or empty.');
        this._name = name;
    }

    updateDescription(description: string): void {
        if (!description) throw new Error('Description cannot be null or empty.');
        this._description = description;
    }

    updateMaxNumberOfRows(maxNumberOfRows: number): void {
        if (maxNumberOfRows === 0) throw new Error('Max number of rows must be greater than zero.');
        this._maxNumberOfRows = maxNumberOfRows;
    }

    updateMaxNumberOfBays(maxNumberOfBays: number): void {
        if (maxNumberOfBays === 0) throw new Error('Max number of bays must be greater than zero.');
        this._maxNumberOfBays = maxNumberOfBays;
    }

    updateMaxNumberOfTiers(maxNumberOfTiers: number): void {
        if (maxNumberOfTiers === 0) throw new Error('Max number of tiers must be greater than zero.');
        this._maxNumberOfTiers = maxNumberOfTiers;
    }

    updatePhysicalCharacteristics(physicalCharacteristics: PhysicalCharacteristics): void {
        if (!physicalCharacteristics) throw new Error('Physical characteristics cannot be null.');
        this._physicalCharacteristics = physicalCharacteristics;
    }

    toDto(): VesselTypeDto {
        return {
            name: this._name,
            description: this._description,
            maxNumberOfRows: this._maxNumberOfRows,
            maxNumberOfBays: this._maxNumberOfBays,
            maxNumberOfTiers: this._maxNumberOfTiers,
            capacity: this.capacity,
            physicalCharacteristics: this._physicalCharacteristics,
        };
    }
}

